lib/runtime.cpp

```diff
@@ -0,0 +1,228 @@
+                                                          mqtt_server_address, mqtt_server_port);
+
+        // module import
+        EVLOG(info) << fmt::format("Initializing module {}...", module_identifier);
+
+        if (!everest.connect()) {
+            EVLOG(critical) << fmt::format("Cannot connect to MQTT broker at {}:{}", mqtt_server_address,
+                                           mqtt_server_port);
+            return 1;
+        }
+
+        auto call_cmd = [&everest](const Requirement& req, const std::string& cmd_name, Parameters args) {
+            return everest.call_cmd(req, cmd_name, args);
+        };
+        this->callbacks.register_call_handler(call_cmd);
+
+        auto publish_var = [&everest](const std::string& param1, const std::string& param2, Value param3) {
+            return everest.publish_var(param1, param2, param3);
+        };
+        this->callbacks.register_publish_handler(publish_var);
+
+        auto subscribe_var = [&everest](const Requirement& req, const std::string& var_name,
+                                        const ValueCallback& callback) {
+            return everest.subscribe_var(req, var_name, callback);
+        };
+        this->callbacks.register_subscribe_handler(subscribe_var);
+
+        // NOLINTNEXTLINE(modernize-avoid-bind): prefer bind here for readability
+        auto external_mqtt_publish =
+            std::bind(&Everest::Everest::external_mqtt_publish, &everest, std::placeholders::_1, std::placeholders::_2);
+        this->callbacks.register_ext_mqtt_publish_handler(external_mqtt_publish);
+
+        // NOLINTNEXTLINE(modernize-avoid-bind): prefer bind here for readability
+        auto external_mqtt_handler = std::bind(&Everest::Everest::provide_external_mqtt_handler, &everest,
+                                               std::placeholders::_1, std::placeholders::_2);
+        this->callbacks.register_ext_mqtt_subscribe_handler(external_mqtt_handler);
+
+        // FIXME (aw): would be nice to move this config related thing toward the module_init function
+        std::vector<cmd> cmds =
+            this->callbacks.everest_register(config.get_main_config()[this->module_id]["connections"]);
+
+        for (auto const& command : cmds) {
+            everest.provide_cmd(command);
+        }
+
+        auto module_configs = config.get_module_configs(this->module_id);
+        const auto module_info = config.get_module_info(this->module_id);
+
+        this->callbacks.init(module_configs, module_info);
+
+        std::thread mainloop_thread = std::thread(&Everest::Everest::mainloop, &everest);
+
+        // register the modules ready handler with the framework
+        // this handler gets called when the global ready signal is received
+        everest.register_on_ready_handler(this->callbacks.ready);
+
+        // the module should now be ready
+        everest.signal_ready();
+
+        mainloop_thread.join();
+
```

Comment:
I guess there is a lot of code duplication here with main.cpp.  If we will drop main.cpp, then this doesn't matter.  If not, it would be nice to share common source.

Reply:
I think I'll drop the modules as shared libraries completely and remove the main.cpp in that process.